sample_consensus/include/pcl/sample_consensus/impl/sac_model_torus.hpp

```diff
@@ -0,0 +1,518 @@
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointT, typename PointNT>
+bool
+pcl::SampleConsensusModelTorus<PointT, PointNT>::computeModelCoefficients(
+    const Indices& samples, Eigen::VectorXf& model_coefficients) const
+{
+
+  // Make sure that the samples are valid
+  if (!isSampleGood(samples)) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::computeModelCoefficients] Invalid set "
+              "of samples given!\n");
+    return (false);
+  }
+
+  if (!normals_) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::computeModelCoefficients] No input "
+              "dataset containing normals was given!\n");
+    return (false);
+  }
+  // Find axis using:
+
+  // @article{article,
+  // author = {Lukacs, G. and Marshall, David and Martin, R.},
+  // year = {2001},
+  // month = {09},
+  // pages = {},
+  // title = {Geometric Least-Squares Fitting of Spheres, Cylinders, Cones and Tori}
+  //}
+
+  Eigen::Vector3f n0 = Eigen::Vector3f((*normals_)[samples[0]].getNormalVector3fMap());
+  Eigen::Vector3f n1 = Eigen::Vector3f((*normals_)[samples[1]].getNormalVector3fMap());
+  Eigen::Vector3f n2 = Eigen::Vector3f((*normals_)[samples[2]].getNormalVector3fMap());
+  Eigen::Vector3f n3 = Eigen::Vector3f((*normals_)[samples[3]].getNormalVector3fMap());
+
+  Eigen::Vector3f p0 = Eigen::Vector3f((*input_)[samples[0]].getVector3fMap());
+  Eigen::Vector3f p1 = Eigen::Vector3f((*input_)[samples[1]].getVector3fMap());
+  Eigen::Vector3f p2 = Eigen::Vector3f((*input_)[samples[2]].getVector3fMap());
+  Eigen::Vector3f p3 = Eigen::Vector3f((*input_)[samples[3]].getVector3fMap());
+
+  float a01 = crossDot(n0, n1, n2);
+  float b01 = crossDot(n0, n1, n3);
+  float a0 = crossDot(p2 - p1, n0, n2);
+  float a1 = crossDot(p0 - p2, n1, n2);
+  float b0 = crossDot(p3 - p1, n0, n3);
+  float b1 = crossDot(p0 - p3, n1, n3);
+  float a = crossDot(p0 - p2, p1 - p0, n2);
+  float b = crossDot(p0 - p3, p1 - p0, n3);
+
+  // a10*t0*t1 + a0*t0 + a1*t1 + a = 0
+  // b10*t0*t1 + b0*t0 + b1*t1 + b = 0
+  //
+  // (a0 - b0*a10/b10)* t0 + (a1-b1*a10/b10) *t1 + a - b*a10/b10
+  // t0 = k * t1 + p
+
+  float k = -(a1 - b1 * a01 / b01) / (a0 - b0 * a01 / b01);
+  float p = -(a - b * a01 / b01) / (a0 - b0 * a01 / b01);
+
+  // Second deg eqn.
+  //
+  // b10*k*t1*t1 + b10*p*t1  | + b0*k *t1 + b0*p | + b1*t1 | + b = 0
+  //
+  // (b10*k) * t1*t1 + (b10*p + b0*k + b1) * t1  + (b0*p + b)
+
+  float _a = (b01 * k);
+  float _b = (b01 * p + b0 * k + b1);
+  float _c = (b0 * p + b);
+
+  float eps = 1e-9; //TODO, what is the right way in pcl?
+  // Check for imaginary solutions, or small denominators.
+  if ((_b * _b - 4 * _a * _c) < 0 || std::abs(a0 - b0 * a01) < eps ||
+      std::abs(b01) < eps || std::abs(_a) < eps) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::computeModelCoefficients] Can't "
+              "compute model coefficients with this method!\n");
```

Comment:
well, technically, it could still work with the refined _pure ML _ method. I still need to confirm this but i think the method assumes that point normals will cross the direction line of the torus, its not the case in the poles of the torus (the locus of this is a circle at the bottom and on the top. I agree that this might not be an actual proper message. I has just trying to cover all possible numerical stability cases.

Reply:
Let's perhaps change this to `PCL_DEBUG` to hide the message, so this does not confuse anyone.